<a href="https://colab.research.google.com/github/Abhilitcode/Next_Word_Prediction_using_LSTM/blob/main/Next_Word_Prediction_using_LSTM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
faqs = """About the Program
What is the course fee for  Data Science Mentorship Program (DSMP 2023)
The course follows a monthly subscription model where you have to make monthly payments of Rs 799/month.
What is the total duration of the course?
The total duration of the course is 7 months. So the total course fee becomes 799*7 = Rs 5600(approx.)
What is the syllabus of the mentorship program?
We will be covering the following modules:
Python Fundamentals
Python libraries for Data Science
Data Analysis
SQL for Data Science
Maths for Machine Learning
ML Algorithms
Practical ML
MLOPs
Case studies
You can check the detailed syllabus here - https://learnwith.campusx.in/courses/CampusX-Data-Science-Mentorship-Program-637339afe4b0615a1bbed390
Will Deep Learning and NLP be a part of this program?
No, NLP and Deep Learning both are not a part of this program’s curriculum.
What if I miss a live session? Will I get a recording of the session?
Yes all our sessions are recorded, so even if you miss a session you can go back and watch the recording.
Where can I find the class schedule?
Checkout this google sheet to see month by month time table of the course - https://docs.google.com/spreadsheets/d/16OoTax_A6ORAeCg4emgexhqqPv3noQPYKU7RJ6ArOzk/edit?usp=sharing.
What is the time duration of all the live sessions?
Roughly, all the sessions last 2 hours.
What is the language spoken by the instructor during the sessions?
Hinglish
How will I be informed about the upcoming class?
You will get a mail from our side before every paid session once you become a paid user.
Can I do this course if I am from a non-tech background?
Yes, absolutely.
I am late, can I join the program in the middle?
Absolutely, you can join the program anytime.
If I join/pay in the middle, will I be able to see all the past lectures?
Yes, once you make the payment you will be able to see all the past content in your dashboard.
Where do I have to submit the task?
You don’t have to submit the task. We will provide you with the solutions, you have to self evaluate the task yourself.
Will we do case studies in the program?
Yes.
Where can we contact you?
You can mail us at nitish.campusx@gmail.com
Payment/Registration related questions
Where do we have to make our payments? Your YouTube channel or website?
You have to make all your monthly payments on our website. Here is the link for our website - https://learnwith.campusx.in/
Can we pay the entire amount of Rs 5600 all at once?
Unfortunately no, the program follows a monthly subscription model.
What is the validity of monthly subscription? Suppose if I pay on 15th Jan, then do I have to pay again on 1st Feb or 15th Feb
15th Feb. The validity period is 30 days from the day you make the payment. So essentially you can join anytime you don’t have to wait for a month to end.
What if I don’t like the course after making the payment. What is the refund policy?
You get a 7 days refund period from the day you have made the payment.
I am living outside India and I am not able to make the payment on the website, what should I do?
You have to contact us by sending a mail at nitish.campusx@gmail.com
Post registration queries
Till when can I view the paid videos on the website?
This one is tricky, so read carefully. You can watch the videos till your subscription is valid. Suppose you have purchased subscription on 21st Jan, you will be able to watch all the past paid sessions in the period of 21st Jan to 20th Feb. But after 21st Feb you will have to purchase the subscription again.
But once the course is over and you have paid us Rs 5600(or 7 installments of Rs 799) you will be able to watch the paid sessions till Aug 2024.
Why lifetime validity is not provided?
Because of the low course fee.
Where can I reach out in case of a doubt after the session?
You will have to fill a google form provided in your dashboard and our team will contact you for a 1 on 1 doubt clearance session
If I join the program late, can I still ask past week doubts?
Yes, just select past week doubt in the doubt clearance google form.
I am living outside India and I am not able to make the payment on the website, what should I do?
You have to contact us by sending a mail at nitish.campusx@gmai.com
Certificate and Placement Assistance related queries
What is the criteria to get the certificate?
There are 2 criterias:
You have to pay the entire fee of Rs 5600
You have to attempt all the course assessments.
I am joining late. How can I pay payment of the earlier months?
You will get a link to pay fee of earlier months in your dashboard once you pay for the current month.
I have read that Placement assistance is a part of this program. What comes under Placement assistance?
This is to clarify that Placement assistance does not mean Placement guarantee. So we dont guarantee you any jobs or for that matter even interview calls. So if you are planning to join this course just for placements, I am afraid you will be disappointed. Here is what comes under placement assistance
Portfolio Building sessions
Soft skill sessions
Sessions with industry mentors
Discussion on Job hunting strategies
"""

In [2]:
import tensorflow as tf

In [3]:
from tensorflow.keras.preprocessing.text import Tokenizer

In [4]:
tokenizer = Tokenizer()

You send multiple texts in the input so it should be in the form of lists.

In [5]:
tokenizer.fit_on_texts([faqs])

we can here check that which number is assigned to whcih texts.

In [6]:
tokenizer.word_index

{'the': 1,
 'you': 2,
 'i': 3,
 'to': 4,
 'a': 5,
 'of': 6,
 'is': 7,
 'have': 8,
 'will': 9,
 'can': 10,
 'what': 11,
 'course': 12,
 'program': 13,
 'in': 14,
 'for': 15,
 'all': 16,
 'sessions': 17,
 'on': 18,
 'be': 19,
 'and': 20,
 'this': 21,
 'if': 22,
 'am': 23,
 'pay': 24,
 'payment': 25,
 'make': 26,
 'we': 27,
 'do': 28,
 'subscription': 29,
 'where': 30,
 'rs': 31,
 'so': 32,
 'campusx': 33,
 'session': 34,
 'our': 35,
 'paid': 36,
 'join': 37,
 'able': 38,
 'your': 39,
 'website': 40,
 'placement': 41,
 'fee': 42,
 'data': 43,
 'monthly': 44,
 'month': 45,
 'not': 46,
 'get': 47,
 'yes': 48,
 'once': 49,
 'past': 50,
 'feb': 51,
 'assistance': 52,
 'science': 53,
 '7': 54,
 '5600': 55,
 'are': 56,
 'watch': 57,
 'google': 58,
 'by': 59,
 'com': 60,
 'mail': 61,
 'from': 62,
 'contact': 63,
 'us': 64,
 'at': 65,
 'or': 66,
 'doubt': 67,
 'mentorship': 68,
 'payments': 69,
 '799': 70,
 'total': 71,
 'duration': 72,
 'months': 73,
 'learning': 74,
 'case': 75,
 'here': 76,
 '

Split the dataset on the basis of the sentences .

In [7]:
for sentences in faqs.split('\n'):
  print(sentences)


About the Program
What is the course fee for  Data Science Mentorship Program (DSMP 2023)
The course follows a monthly subscription model where you have to make monthly payments of Rs 799/month.
What is the total duration of the course?
The total duration of the course is 7 months. So the total course fee becomes 799*7 = Rs 5600(approx.)
What is the syllabus of the mentorship program?
We will be covering the following modules:
Python Fundamentals
Python libraries for Data Science
Data Analysis
SQL for Data Science
Maths for Machine Learning
ML Algorithms
Practical ML
MLOPs
Case studies
You can check the detailed syllabus here - https://learnwith.campusx.in/courses/CampusX-Data-Science-Mentorship-Program-637339afe4b0615a1bbed390
Will Deep Learning and NLP be a part of this program?
No, NLP and Deep Learning both are not a part of this program’s curriculum.
What if I miss a live session? Will I get a recording of the session?
Yes all our sessions are recorded, so even if you miss a sessi

Now, when you use texts_to_sequences, it will convert the words in your texts into their respective indices:

The reason you get **lists of lists** when you use `tokenizer.texts_to_sequences` is that the method is designed to handle **multiple texts** at once, even if you provide just a single text. Here's a detailed explanation:

### Why You Get Lists of Lists:
- `texts_to_sequences` expects a **list of texts** as input. It processes each text and returns a list where each item is a list of tokenized words for the corresponding input text.
- Even when you provide a single text wrapped in a list (e.g., `texts_to_sequences([sentence])`), it still processes that input as a list of one item and returns the result as a list containing one sequence.

### Example:
Let’s break down a simple case:

#### 1. **Multiple Sentences Input**:
When you give it multiple sentences, the method will return a list of sequences, one for each sentence:

```python
texts = ["I love programming", "AI is amazing"]

tokenizer = Tokenizer()
tokenizer.fit_on_texts(texts)

sequences = tokenizer.texts_to_sequences(texts)
print(sequences)
```

**Output**:
```python
[[1, 2, 3], [4, 5, 6]]
```

Here’s what happens:
- The first sentence, "I love programming", is converted into `[1, 2, 3]`.
- The second sentence, "AI is amazing", is converted into `[4, 5, 6]`.

Since you provided a list of two texts, it returns a **list of two lists**, each representing the tokenized sequence of the corresponding sentence.

#### 2. **Single Sentence Input**:
Even if you only pass a single sentence wrapped in a list, `texts_to_sequences` still returns a list of sequences, but in this case, it's a list containing just one list:

```python
text = ["I love programming"]

sequences = tokenizer.texts_to_sequences(text)
print(sequences)
```

**Output**:
```python
[[1, 2, 3]]
```

- Even though there is only one sentence, the method returns a list of lists because it expects to handle multiple texts. Here, it's a list with one item: `[[1, 2, 3]]`.

### When You Only Want One Sequence:
If you're only processing one text at a time, you can use `[0]` to extract the inner list:

```python
print(sequences[0])  # Output: [1, 2, 3]
```

### Summary:
- **`texts_to_sequences` returns lists of lists** because it’s designed to handle multiple sentences at once. Each text is tokenized into a sequence of integers, and those sequences are returned as a list of lists.
- If you're only passing a single text and want just the sequence, you can access the first (and only) list using `[0]`.

This behavior is typical of functions in libraries like TensorFlow/Keras, which are designed to process **batches of data** (like multiple sentences or texts) efficiently.

Let me know if that clears things up!

In [47]:
for sentences in faqs.split('\n'):
  print(tokenizer.texts_to_sequences([sentences])[0])

[93, 1, 13]
[11, 7, 1, 12, 42, 15, 43, 53, 68, 13, 147, 148]
[1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6, 31, 70, 45]
[11, 7, 1, 71, 72, 6, 1, 12]
[1, 71, 72, 6, 1, 12, 7, 54, 73, 32, 1, 71, 12, 42, 149, 70, 54, 31, 55, 150]
[11, 7, 1, 96, 6, 1, 68, 13]
[27, 9, 19, 151, 1, 152, 153]
[97, 154]
[97, 155, 15, 43, 53]
[43, 156]
[157, 15, 43, 53]
[158, 15, 159, 74]
[98, 160]
[161, 98]
[162]
[75, 99]
[2, 10, 163, 1, 164, 96, 76, 77, 100, 33, 14, 165, 33, 43, 53, 68, 13, 166]
[9, 101, 74, 20, 102, 19, 5, 78, 6, 21, 13]
[103, 102, 20, 101, 74, 167, 56, 46, 5, 78, 6, 21, 168, 169]
[11, 22, 3, 104, 5, 105, 34, 9, 3, 47, 5, 106, 6, 1, 34]
[48, 16, 35, 17, 56, 170, 32, 107, 22, 2, 104, 5, 34, 2, 10, 171, 172, 20, 57, 1, 106]
[30, 10, 3, 173, 1, 108, 174]
[175, 21, 58, 176, 4, 79, 45, 59, 45, 109, 177, 6, 1, 12, 77, 178, 58, 60, 179, 180, 181, 182, 183, 184, 185]
[11, 7, 1, 109, 72, 6, 16, 1, 105, 17]
[186, 16, 1, 17, 187, 110, 188]
[11, 7, 1, 189, 190, 59, 1, 191, 192, 1, 17]
[193]
[111,

In [8]:
list_of_sentences = []
for sentences in faqs.split('\n'):
  tokenized_sentences = tokenizer.texts_to_sequences([sentences])[0]
  for i in range(1,len(tokenized_sentences)):
    list_of_sentences.append(tokenized_sentences[:i+1])


These are only the input sequences not yet extracted the output

In [9]:
list_of_sentences

[[93, 1],
 [93, 1, 13],
 [11, 7],
 [11, 7, 1],
 [11, 7, 1, 12],
 [11, 7, 1, 12, 42],
 [11, 7, 1, 12, 42, 15],
 [11, 7, 1, 12, 42, 15, 43],
 [11, 7, 1, 12, 42, 15, 43, 53],
 [11, 7, 1, 12, 42, 15, 43, 53, 68],
 [11, 7, 1, 12, 42, 15, 43, 53, 68, 13],
 [11, 7, 1, 12, 42, 15, 43, 53, 68, 13, 147],
 [11, 7, 1, 12, 42, 15, 43, 53, 68, 13, 147, 148],
 [1, 12],
 [1, 12, 94],
 [1, 12, 94, 5],
 [1, 12, 94, 5, 44],
 [1, 12, 94, 5, 44, 29],
 [1, 12, 94, 5, 44, 29, 95],
 [1, 12, 94, 5, 44, 29, 95, 30],
 [1, 12, 94, 5, 44, 29, 95, 30, 2],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6, 31],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6, 31, 70],
 [1, 12, 94, 5, 44, 29, 95, 30, 2, 8, 4, 26, 44, 69, 6

Padding is needed in sequence-based models, such as those used in Natural Language Processing (NLP), because most models (e.g., RNNs, LSTMs, Transformers) expect input sequences to be of the **same length** for efficient batch processing. Here's why padding is necessary and how it works in your case.

### Why Padding is Necessary:
1. **Fixed-Length Inputs**: Most deep learning models, especially when training in batches, expect input sequences to have the same length. If sequences have different lengths, the model won’t be able to process them uniformly in one batch.
  
2. **Variable-Length Sequences**: In real-world data, text sequences (sentences) can vary in length. Some sentences might be very short, while others can be long. In your example, the sequences you're generating will likely have different lengths.
   
   Example of tokenized sequences:
   ```python
   [[1, 2],
    [1, 2, 3, 4],
    [1, 2, 3],
    [1]]
   ```
   These sequences cannot be processed together without padding because they don’t have the same number of tokens.

3. **Alignment**: Padding ensures all sequences in a batch are aligned to the same length. The shorter sequences are padded with a special token (usually `0`), which indicates that those positions should not be considered during training.

### Example:

Suppose your tokenized sequences look like this:

```python
list_of_sentences = [[1, 2], [1, 2, 3, 4], [1, 2, 3], [1]]
```

Without padding, the sequences have different lengths (2, 4, 3, and 1 tokens). If you pass this directly to a model, it will throw an error due to mismatched lengths.

By applying padding, the sequences will be padded to the same length, like this:

```python
padded_sequences = [[1, 2, 0, 0],
                    [1, 2, 3, 4],
                    [1, 2, 3, 0],
                    [1, 0, 0, 0]]
```

### How Padding Works:
- Padding adds extra tokens (usually `0`s) to shorter sequences so that all sequences in a batch have the same length. The length is typically equal to the longest sequence in the batch.
- Models usually treat these `0` tokens as "no information" and ignore them during the learning process.

### Keras Example (How to Apply Padding):
In Keras, you can use the `pad_sequences` function to handle padding automatically:

```python
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Assuming list_of_sentences contains your tokenized sequences
padded_sequences = pad_sequences(list_of_sentences, padding='post')

print(padded_sequences)
```

This will pad the sequences to the length of the longest sequence in your list.

**Output**:
```python
[[1, 2, 0, 0],
 [1, 2, 3, 4],
 [1, 2, 3, 0],
 [1, 0, 0, 0]]
```

### Types of Padding:
- **Pre-Padding (`padding='pre'`)**: Pads with zeros at the beginning of the sequence.
  Example: `[[0, 0, 1, 2], [1, 2, 3, 4]]`
  
- **Post-Padding (`padding='post'`)**: Pads with zeros at the end of the sequence (this is generally more common).
  Example: `[[1, 2, 0, 0], [1, 2, 3, 4]]`

### Why Padding is Needed in Your Code:
In your case, you’re appending incrementally longer sub-sequences to `list_of_sentences`. This means that sequences of different lengths are being generated. Padding ensures that when these sequences are passed to a machine learning model, they can be processed uniformly.

Without padding, sequences of different lengths can't be processed together in a single batch, making training inefficient or even impossible in some cases.

### Summary:
- **Padding** ensures that all sequences are of the same length, which is required by most models that work with batches of data.
- It is necessary because your sequences will naturally have different lengths, and without padding, the model cannot handle them efficiently.
- `pad_sequences` can be used to automatically handle the padding, aligning all sequences to the same length by adding zeros.


In [10]:
len_inp_seq = [len(x) for x in list_of_sentences]

In [11]:
len_inp_seq

[2,
 3,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 2,
 3,
 4,
 5,
 6,
 7,
 2,
 2,
 3,
 4,
 5,
 2,
 2,
 3,
 4,
 2,
 3,
 4,
 2,
 2,
 2,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 2,
 3,
 4,
 5,
 6,
 7,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 2,
 3,
 4,
 5,
 6,
 7,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11

In [12]:
max_len_seq = max([len(x) for x in list_of_sentences])

In [13]:
max_len_seq

57

Apply padding here.

The `maxlen` parameter in the `pad_sequences` function controls the **maximum length** of the sequences after padding (or truncation). It allows you to specify a fixed length for all the sequences, regardless of their original lengths.

### How `maxlen` Works:

- If a sequence is **shorter** than `maxlen`, padding is added to reach the desired length.
- If a sequence is **longer** than `maxlen`, it gets **truncated** (cut down) to the specified length.
- You can control whether the padding or truncation happens **before** or **after** the sequence using the `padding` and `truncating` parameters.

### Example Usage:

```python
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Example: list of tokenized sentences (variable length)
sequences = [[1, 2, 3], [4, 5], [6, 7, 8, 9, 10]]

# Applying padding with a specific maxlen
padded_sequences = pad_sequences(sequences, maxlen=3, padding='post', truncating='post')

print(padded_sequences)
```

**Output**:
```python
[[ 1  2  3]
 [ 4  5  0]
 [ 8  9 10]]
```

### Explanation:
- The `maxlen=3` parameter ensures that each sequence is **exactly 3 tokens long**.
- Sequences that are **shorter than 3** (like `[4, 5]`) get padded with zeros (`[4, 5, 0]`).
- Sequences that are **longer than 3** (like `[6, 7, 8, 9, 10]`) get truncated (`[8, 9, 10]`).

### Key Parameters with `maxlen`:

- **`maxlen`**: Defines the maximum length of all sequences. Any sequences longer than this will be truncated, and shorter ones will be padded.
  
- **`padding`**: Controls where padding is added.
  - `'pre'` (default): Adds padding **before** the sequence.
  - `'post'`: Adds padding **after** the sequence.

- **`truncating`**: Controls where truncation happens.
  - `'pre'`: Truncates **before** the sequence.
  - `'post'` (default): Truncates **after** the sequence.

### Example With `maxlen=5`:

```python
padded_sequences = pad_sequences(sequences, maxlen=5, padding='pre', truncating='post')

print(padded_sequences)
```

**Output**:
```python
[[ 0  0  1  2  3]
 [ 0  0  0  4  5]
 [ 6  7  8  9 10]]
```

Here, the sequences are padded with zeros before (since `padding='pre'`), and truncation happens at the end if the sequence is longer than 5 (which didn't happen in this case because all sequences are shorter than or equal to 5).

### When Should You Use `maxlen`?

- **To Control Sequence Lengths**: If you want all your sequences to have the same length for uniform processing, but you don’t want extremely long sequences, you can set `maxlen` to limit them.
- **To Prevent Memory Issues**: Truncating overly long sequences can help manage memory usage and avoid inefficiency during training, especially when dealing with long texts.
- **To Focus on the Most Important Tokens**: In tasks like sentiment analysis or text classification, you might only need the most recent tokens (truncate earlier tokens) or the first few tokens (truncate later tokens) to capture the key information.

### Summary:
- `maxlen` sets the **maximum allowed length** for sequences.
- If sequences are **longer**, they get truncated. If they are **shorter**, they get padded to match the `maxlen`.
- You can control **where** the padding or truncation happens using `padding` and `truncating`.


In [16]:
from tensorflow.keras.preprocessing.sequence import pad_sequences

In [17]:
padded_sequences = pad_sequences(list_of_sentences,maxlen = max_len_seq, padding='pre')

In [18]:
padded_sequences

array([[  0,   0,   0, ...,   0,  93,   1],
       [  0,   0,   0, ...,  93,   1,  13],
       [  0,   0,   0, ...,   0,  11,   7],
       ...,
       [  0,   0,   0, ..., 279,  18, 280],
       [  0,   0,   0, ...,  18, 280, 281],
       [  0,   0,   0, ..., 280, 281, 282]], dtype=int32)

Split them in X and Y. X will have inputs and output will be the last value. of sequences.

In the line of code you provided:

```python
X = padded_sequences[:, :-1]
```

This is using **slicing** on `padded_sequences` to create a new array `X` by **removing the last element from each sequence**. Let's break it down step by step:

### Explanation of `padded_sequences[:, :-1]`:

1. **`padded_sequences`**:
   - This is a 2D array (a list of lists) containing sequences of tokens, where each sequence has the same length due to padding.

2. **`:` (colon)**:
   - The colon on its own means "select all rows". So this part selects all sequences (rows) from `padded_sequences`.

3. **`:-1`**:
   - This means "select all elements except the last one" for each row (sequence). In Python slicing, negative indices count from the end, so `-1` refers to the last element.
   - `:-1` means "from the start up to (but not including) the last element".

### What This Code Does:
It takes the 2D `padded_sequences` array and returns a new array `X` where **each sequence has its last token removed**.

### Example:

Suppose `padded_sequences` looks like this (with padding):

```python
padded_sequences = [[1, 2, 3, 0],
                    [4, 5, 6, 0],
                    [7, 8, 9, 0]]
```

Now, applying the code:

```python
X = padded_sequences[:, :-1]
```

`X` will be:

```python
X = [[1, 2, 3],
     [4, 5, 6],
     [7, 8, 9]]
```

Here, the last padded element `0` (or any other last token) is removed from each sequence.

### Why You Might Do This:
1. **Input to the Model**: In many NLP tasks, you might need to remove the last token from the input sequences before feeding them into the model, especially in **sequence-to-sequence** models (like machine translation or text generation).
   - You often train models to predict the next word based on the current word, so the last token is removed from the input because there’s no "next word" for it.

2. **Target Generation**: If you are using `padded_sequences` to generate input (`X`) and target sequences (`y`), you might remove the last token from the input `X` and use it as part of the output sequence `y`. For example:
   - Input (`X`): `[[1, 2], [4, 5]]`
   - Target (`y`): `[[2, 3], [5, 6]]`

This approach aligns well with tasks where the model learns to predict the next token in a sequence (like language models).

### Summary:
- `X = padded_sequences[:, :-1]` selects all rows from `padded_sequences` and removes the last element from each sequence.
- This is useful when you need to prepare input sequences where the last token is irrelevant for prediction, such as in sequence-to-sequence tasks or language modeling.


In [19]:
X = padded_sequences[:,:-1]

In [20]:
X

array([[  0,   0,   0, ...,   0,   0,  93],
       [  0,   0,   0, ...,   0,  93,   1],
       [  0,   0,   0, ...,   0,   0,  11],
       ...,
       [  0,   0,   0, ...,   0, 279,  18],
       [  0,   0,   0, ..., 279,  18, 280],
       [  0,   0,   0, ...,  18, 280, 281]], dtype=int32)

The line of code:

```python
y = padded_input_sequences[:, -1]
```

is extracting the **last element from each sequence** in `padded_input_sequences`. This is typically done to prepare the **target labels** (`y`) for a model when working with sequence-based tasks like sequence-to-sequence modeling, next-word prediction, or language modeling.

### Explanation of `padded_input_sequences[:, -1]`:

1. **`padded_input_sequences`**:
   - This is a 2D array (a list of lists) containing sequences of tokens, where each sequence is padded to have the same length.

2. **`:` (colon)**:
   - The colon means "select all rows". So, this selects **all sequences** from the input data.

3. **`-1`**:
   - `-1` refers to the **last element** in each sequence. In Python, negative indices start from the end, so `-1` is the last item in each row (sequence).

### What This Code Does:
It creates a new array `y` that contains the **last token of each sequence** from `padded_input_sequences`. This is typically used when the last token represents the **target** you want to predict.

### Example:

Suppose `padded_input_sequences` looks like this:

```python
padded_input_sequences = [[1, 2, 3, 4],
                          [5, 6, 7, 8],
                          [9, 10, 11, 12]]
```

Now, applying the code:

```python
y = padded_input_sequences[:, -1]
```

`y` will be:

```python
y = [4, 8, 12]
```

In this example:
- From the first sequence `[1, 2, 3, 4]`, it extracts `4`.
- From the second sequence `[5, 6, 7, 8]`, it extracts `8`.
- From the third sequence `[9, 10, 11, 12]`, it extracts `12`.

### Why You Might Do This:
1. **Next Word Prediction**: If you are training a model to predict the next token in a sequence (e.g., language modeling or text generation), the last token in each sequence (`y`) can be the **target label** that the model should predict.
   
   - For example, if the input is `X = [1, 2, 3]`, then the model's task might be to predict the next token `y = 4`.

2. **Supervised Learning**: In supervised learning tasks, the last token might represent the label or the output you're trying to predict based on the preceding tokens. The tokens before the last one are the **input**, and the last token is the **target output**.

### Pairing `X` and `y`:
If you're using `X = padded_sequences[:, :-1]` and `y = padded_input_sequences[:, -1]`, here's how it works together:

- **`X`**: The input sequences **without the last token** (for prediction).
- **`y`**: The **last token** in each sequence, which serves as the label or the target.

### Example Workflow:
Given the input sequences:

```python
padded_input_sequences = [[1, 2, 3, 4],
                          [5, 6, 7, 8],
                          [9, 10, 11, 12]]
```

1. **Input for the model (`X`)**: Removing the last token:
   ```python
   X = padded_input_sequences[:, :-1]
   print(X)
   # Output: [[1, 2, 3], [5, 6, 7], [9, 10, 11]]
   ```

2. **Target for the model (`y`)**: Extracting the last token:
   ```python
   y = padded_input_sequences[:, -1]
   print(y)
   # Output: [4, 8, 12]
   ```

This setup can be useful in tasks like **sequence prediction**, where the model learns to predict the next token in a sequence.

### Summary:
- `y = padded_input_sequences[:, -1]` extracts the **last token** from each sequence, often used as the **target label** in sequence-based tasks.
- It's paired with `X = padded_input_sequences[:, :-1]`, where the model is given the preceding tokens (`X`) and trained to predict the last token (`y`).


In [21]:
y = padded_sequences[:,-1]

In [22]:
y

array([  1,  13,   7,   1,  12,  42,  15,  43,  53,  68,  13, 147, 148,
        12,  94,   5,  44,  29,  95,  30,   2,   8,   4,  26,  44,  69,
         6,  31,  70,  45,   7,   1,  71,  72,   6,   1,  12,  71,  72,
         6,   1,  12,   7,  54,  73,  32,   1,  71,  12,  42, 149,  70,
        54,  31,  55, 150,   7,   1,  96,   6,   1,  68,  13,   9,  19,
       151,   1, 152, 153, 154, 155,  15,  43,  53, 156,  15,  43,  53,
        15, 159,  74, 160,  98,  99,  10, 163,   1, 164,  96,  76,  77,
       100,  33,  14, 165,  33,  43,  53,  68,  13, 166, 101,  74,  20,
       102,  19,   5,  78,   6,  21,  13, 102,  20, 101,  74, 167,  56,
        46,   5,  78,   6,  21, 168, 169,  22,   3, 104,   5, 105,  34,
         9,   3,  47,   5, 106,   6,   1,  34,  16,  35,  17,  56, 170,
        32, 107,  22,   2, 104,   5,  34,   2,  10, 171, 172,  20,  57,
         1, 106,  10,   3, 173,   1, 108, 174,  21,  58, 176,   4,  79,
        45,  59,  45, 109, 177,   6,   1,  12,  77, 178,  58,  6

In [23]:
X.shape

(863, 56)

In [24]:
y.shape

(863,)

We now have achieved our dataset that is input:ouput. basically input mapped with its output. So we converted a text gen -> supervised learning so will it be a regression problem or classification?

So ans is classification. if we consider it as regression than we can get output such as 2.1,3.1 here no words are associated with such numbers. So the reason it is a classification problem.

we will apply one hot encoding to the o/p varibles to convert them from scaler to vector. also num_classes = no.of words. always add a count of 1 word extra beacuse in one hot encoding he index starts from 0. and here we have sequences starting from 1 so if we use 282 only words it will miss the last word.

In [25]:
from tensorflow.keras.utils import to_categorical

In [26]:
len(tokenizer.word_index)

282

In [27]:
y = to_categorical(y, num_classes=283)

863 refers to the total number of samples (or sequences) in your dataset. Each sample in x has 56 features, and each corresponding label in y represents one of the 282 unique words (or a one-hot encoded vector of size 283 after encoding).

In [28]:
y.shape

(863, 283)

In [29]:
y

array([[0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

This is a sequential model built using Keras for a task that likely involves text classification or sequence prediction. Let’s break down each layer and explain its purpose in detail:

### 1. **Embedding Layer**
```python
model.add(Embedding(283, 100, input_length=56))
```
- **Embedding(283, 100)**:
  - The **embedding layer** converts each word (represented as an integer) into a dense vector of fixed size (100 in this case).
  - **283**: The vocabulary size, i.e., the total number of unique words in your tokenizer (based on the length of `tokenizer.word_index + 1`).
  - **100**: The dimension of the word embeddings. This means each word will be represented by a vector of size 100. These vectors capture the semantic meaning of words.
  
- **input_length=56**:
  - The input sequences are all of length **56** (each sequence is padded/truncated to 56 tokens). This tells the model the shape of the input data, where each sequence contains 56 words.

**Why?**
- The embedding layer allows the model to learn dense, continuous representations of words in a lower-dimensional space (instead of just one-hot encoded vectors). It helps capture word semantics and relationships between words.

---

### 2. **First LSTM Layer**
```python
model.add(LSTM(150))
```
- **LSTM(150)**:
  - An LSTM (Long Short-Term Memory) layer with **150 units** (or neurons). LSTMs are a type of recurrent neural network (RNN) designed to handle sequence data, especially for capturing long-range dependencies and patterns in the data.
  
- This layer processes the embedded input sequences. It can remember important information from the past and decide which information to retain or discard.

**Why?**
- LSTMs are well-suited for sequence-based tasks, like text, as they can capture temporal dependencies (e.g., how the words in a sentence relate to one another).

---

### 3. **Second LSTM Layer**
```python
model.add(LSTM(150))
```
- Another LSTM layer with **150 units** is added. The idea here is to stack LSTM layers, which often helps the model learn more complex features from the data by adding more depth.
  
- **Why stack LSTMs?**
  - Stacking LSTM layers allows the model to learn more abstract representations of the sequence at each level. The first LSTM might capture simpler patterns (e.g., word-level dependencies), and the second LSTM can capture higher-level patterns (e.g., sentence-level dependencies).

---

### 4. **Dense (Fully Connected) Layer**
```python
model.add(Dense(283, activation='softmax'))
```
- **Dense(283)**:
  - A fully connected (dense) layer with **283 output units**. This corresponds to the number of unique classes (words or tokens), which matches the vocabulary size (283).

- **activation='softmax'**:
  - The **softmax** activation function is applied to the output, which converts the outputs into probability distributions. The model will output a probability for each of the 283 possible words.

**Why?**
- The dense layer with softmax activation is used for classification. In this case, the model is likely predicting the next word (or a sequence of words), and the output is a probability distribution over the entire vocabulary.

---

### Why This Architecture?
- **Embedding Layer**: Transforms the input words into dense, continuous vectors that capture semantic meaning.
- **LSTM Layers**: Process the sequences and capture dependencies between words/tokens. Stacking LSTMs helps the model learn complex features at different abstraction levels.
- **Dense Softmax Layer**: Classifies each sequence into one of 283 classes (words) by outputting a probability for each word.

This architecture is typically used for tasks like text generation, machine translation, or sequence-to-sequence modeling, where the model needs to predict the next word in a sequence.

In [30]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, Dense

In [31]:
model = Sequential()

In [34]:
model.add(Embedding(283, 100))
# model.add(LSTM(150))
model.add(LSTM(150))
model.add(Dense(283, activation='softmax'))

The `model.compile` step in Keras is crucial because it configures the model for training. Let’s break down the purpose and key arguments of `model.compile()`:

### Purpose of `model.compile()`
- **Compilation** is the step where you specify the optimization strategy, the loss function, and evaluation metrics for the model. It prepares the model for training by determining how the model will be optimized, how errors will be measured, and what metrics will be tracked.

### Key Arguments in `model.compile()`

1. **`optimizer`**:
   - This defines the optimization algorithm the model will use to update the weights and minimize the loss.
   - Popular optimizers include:
     - `adam`: Adaptive Moment Estimation, which adjusts the learning rate during training based on the first and second moments of the gradient. It's widely used because it works well in many scenarios.
     - `sgd`: Stochastic Gradient Descent, where weights are updated after each training example, and it can converge to a global minimum.
     - Other options: `rmsprop`, `adagrad`, etc.
   
   **Why important?**
   - The optimizer controls how the model learns by adjusting the weights based on the loss function. Different optimizers handle this process in different ways, and choosing the right one can significantly affect training speed and performance.

2. **`loss`**:
   - This is the loss function that measures how well the model performs by comparing its predictions to the true labels. The model tries to minimize this value during training.
   - Common loss functions include:
     - `categorical_crossentropy`: For multi-class classification tasks, used when the target output is one of many classes (like in your example where the output is a word from a vocabulary).
     - `binary_crossentropy`: For binary classification problems.
     - `mean_squared_error`: For regression problems.
   
   **Why important?**
   - The loss function defines what the model is optimizing. It measures the "error" or "difference" between predicted and actual values, guiding the optimizer to minimize that error during training.

3. **`metrics`**:
   - This specifies the metrics that will be used to evaluate the model during training and testing. Metrics like `accuracy` are often used in classification tasks.
   - Example:
     ```python
     metrics=['accuracy']
     ```
   
   **Why important?**
   - Metrics help you monitor the model’s performance. For example, accuracy tells you how often the model's predictions are correct, allowing you to track how well the model generalizes to the data.

### Example of `model.compile()`

```python
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
```

- **`optimizer='adam'`**: Uses the Adam optimizer, which is adaptive and performs well for most tasks.
- **`loss='categorical_crossentropy'`**: Measures the error between the predicted class probabilities and the true class (used for multi-class classification).
- **`metrics=['accuracy']`**: Tracks the accuracy of the model during training and evaluation.

### Why is `model.compile()` necessary?
Without compiling, the model doesn’t know:
- How to update its weights (no optimizer).
- How to measure its performance (no loss function).
- What metric to track (e.g., accuracy).

Thus, `model.compile()` configures these fundamental aspects, making it a necessary step before training your model with `model.fit()`.

In [36]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [37]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding (Embedding)                │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lstm (LSTM)                          │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lstm_1 (LSTM)                        │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ embedding_2 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ lstm_2 (LSTM)                        │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)